Investigate why Forest occasionally exports invalid snapshots

**Issue summary**
<!-- A clear and concise description of what the task is. -->
Forest occasionally exports a snapshot that contain broken references. We need to investigate why this happens.

Tasks:
- [ ] Identify how often snapshots are broken. Could generate/import snapshots in a loop until something breaks.
- [ ] Investigate if it is related to new blocks being received to the HEAD tipset. This can be tested by generating snapshots for epochs that are not HEAD. That is, if HEAD is epoch 1000, generate snapshots for epoch 900 and test if they work.
- [ ] Use the tool developed in #2325 and provide feedback.

Replicate failure with broken snapshot:
- Download snapshot: https://forest-snapshots.fra1.digitaloceanspaces.com/forest_snapshot_calibnet_2022-12-05_height_95741.car
- Import it in Forest.


**Other information and links**
<!-- Add any other context or screenshots about the issue here. -->



<!-- Thank you 🙏 -->

Update:
forest daemon stuck with bad snapshot, printing no error message.  Tried with lotus, printing blow error messages while loading the same bad snapshot

```console
2022-12-09T18:25:54.918+0800    ERROR   chain   chain/sync.go:949       failed to validate tipset: validating block bafy2bzaceboa4lx225nzf5mf25dr4bjmeaoxghxdggxugup2erjothmd4fzni:
    github.com/filecoin-project/lotus/chain.(*Syncer).ValidateTipSet.func1
        /home/me/git/lotus/chain/sync.go:584
  - 3 errors occurred:
        * block had invalid messages:
    github.com/filecoin-project/lotus/chain/consensus/filcns.(*FilecoinEC).ValidateBlock.func1
        /home/me/git/lotus/chain/consensus/filcns/filecoin.go:134
  - failed to compute tipsettate for {bafy2bzacect4widcroebz4ijcul3qew5hn6xif4durayol4hbtb2dq4ysedk4,bafy2bzaceculwegadcv5pkcxqsfb4dhvg2dskxschgj5vu4ie5xolcxbdwsas,bafy2bzacebwb6m4sgote2ydrrg2emep4mb3fu5k3epmumqaw2vgzvvnnagak6,bafy2bzaceajc5o53p22xr45tkkrrqicp2q6qnmvg5cq2hzzgm773wdwmqhn5u,bafy2bzacechzactcbrsfddnv4q7hroklshedxx673k5d6l3zrbrunlvkdsaig,bafy2bzaceat4dyrflcveuybogucth45s25ya5uv5tjkeysdg4qbe6ey3imngs}:
    github.com/filecoin-project/lotus/chain/consensus/filcns.(*FilecoinEC).checkBlockMessages
        /home/me/git/lotus/chain/consensus/filcns/filecoin.go:462
  - getting block messages for tipset:
    github.com/filecoin-project/lotus/chain/consensus/filcns.(*TipSetExecutor).ExecuteTipSet
        /home/me/git/lotus/chain/consensus/filcns/compute_state.go:306
  - failed to get messages for block:
    github.com/filecoin-project/lotus/chain/store.(*ChainStore).BlockMsgsForTipset
        /home/me/git/lotus/chain/store/messages.go:147
  - failed to load msgmeta (bafy2bzacecb6oqlu7x3pufe6byqvpiwdo3vis2jguwhlr54ch6q665qs4bv7w):
    github.com/filecoin-project/lotus/chain/store.(*ChainStore).ReadMsgMetaCids
        /home/me/git/lotus/chain/store/messages.go:219
  - ipld: could not find bafy2bzacecb6oqlu7x3pufe6byqvpiwdo3vis2jguwhlr54ch6q665qs4bv7w
        * computing base fee:
    github.com/filecoin-project/lotus/chain/consensus/filcns.(*FilecoinEC).ValidateBlock.func3
        /home/me/git/lotus/chain/consensus/filcns/filecoin.go:149
  - error getting messages for: bafy2bzacect4widcroebz4ijcul3qew5hn6xif4durayol4hbtb2dq4ysedk4:
    github.com/filecoin-project/lotus/chain/store.(*ChainStore).ComputeBaseFee
        /home/me/git/lotus/chain/store/basefee.go:65
  - failed to load msgmeta (bafy2bzacecb6oqlu7x3pufe6byqvpiwdo3vis2jguwhlr54ch6q665qs4bv7w):
    github.com/filecoin-project/lotus/chain/store.(*ChainStore).ReadMsgMetaCids
        /home/me/git/lotus/chain/store/messages.go:219
  - ipld: could not find bafy2bzacecb6oqlu7x3pufe6byqvpiwdo3vis2jguwhlr54ch6q665qs4bv7w
        * get tipsetstate(102531, [bafy2bzacect4widcroebz4ijcul3qew5hn6xif4durayol4hbtb2dq4ysedk4 bafy2bzaceculwegadcv5pkcxqsfb4dhvg2dskxschgj5vu4ie5xolcxbdwsas bafy2bzacebwb6m4sgote2ydrrg2emep4mb3fu5k3epmumqaw2vgzvvnnagak6 bafy2bzaceajc5o53p22xr45tkkrrqicp2q6qnmvg5cq2hzzgm773wdwmqhn5u bafy2bzacechzactcbrsfddnv4q7hroklshedxx673k5d6l3zrbrunlvkdsaig bafy2bzaceat4dyrflcveuybogucth45s25ya5uv5tjkeysdg4qbe6ey3imngs]) failed:
    github.com/filecoin-project/lotus/chain/consensus/filcns.(*FilecoinEC).ValidateBlock.func4
        /home/me/git/lotus/chain/consensus/filcns/filecoin.go:170
  - getting block messages for tipset:
    github.com/filecoin-project/lotus/chain/consensus/filcns.(*TipSetExecutor).ExecuteTipSet
        /home/me/git/lotus/chain/consensus/filcns/compute_state.go:306
  - failed to get messages for block:
    github.com/filecoin-project/lotus/chain/store.(*ChainStore).BlockMsgsForTipset
        /home/me/git/lotus/chain/store/messages.go:147
  - failed to load msgmeta (bafy2bzacecb6oqlu7x3pufe6byqvpiwdo3vis2jguwhlr54ch6q665qs4bv7w):
    github.com/filecoin-project/lotus/chain/store.(*ChainStore).ReadMsgMetaCids
        /home/me/git/lotus/chain/store/messages.go:219
  - ipld: could not find bafy2bzacecb6oqlu7x3pufe6byqvpiwdo3vis2jguwhlr54ch6q665qs4bv7w

2022-12-09T18:25:54.918+0800    ERROR   chain   chain/sync_manager.go:254       error during sync in [bafy2bzacecob34qhpbv7ad7apxprmt2nxpn5tkhumactb6p72p3dv4zompiak bafy2bzacebgktfpflnntsjd5zzmzfrxwud2hkq2bto52spzg7rgdoq7puekjo bafy2bzacebumotbcaa4haxya44ojpuu5me7zhhuzoc2z2r6zm2gyf2kmfl5nk bafy2bzaced5vscy3o7qt2n3cbh2livc5zideqhuosfma6fpjbetf5xh7jnxn4 bafy2bzacedb5o7l3yj76ofoy7sl5eew752gtzmsgj3xjw6dj6utbro5rm6jpk]: collectChain failed: collectChain syncMessages: message processing failed: validating block bafy2bzaceboa4lx225nzf5mf25dr4bjmeaoxghxdggxugup2erjothmd4fzni: 3 errors occurred:
        * block had invalid messages:
    github.com/filecoin-project/lotus/chain/consensus/filcns.(*FilecoinEC).ValidateBlock.func1
        /home/me/git/lotus/chain/consensus/filcns/filecoin.go:134
  - failed to compute tipsettate for {bafy2bzacect4widcroebz4ijcul3qew5hn6xif4durayol4hbtb2dq4ysedk4,bafy2bzaceculwegadcv5pkcxqsfb4dhvg2dskxschgj5vu4ie5xolcxbdwsas,bafy2bzacebwb6m4sgote2ydrrg2emep4mb3fu5k3epmumqaw2vgzvvnnagak6,bafy2bzaceajc5o53p22xr45tkkrrqicp2q6qnmvg5cq2hzzgm773wdwmqhn5u,bafy2bzacechzactcbrsfddnv4q7hroklshedxx673k5d6l3zrbrunlvkdsaig,bafy2bzaceat4dyrflcveuybogucth45s25ya5uv5tjkeysdg4qbe6ey3imngs}:
    github.com/filecoin-project/lotus/chain/consensus/filcns.(*FilecoinEC).checkBlockMessages
        /home/me/git/lotus/chain/consensus/filcns/filecoin.go:462
  - getting block messages for tipset:
    github.com/filecoin-project/lotus/chain/consensus/filcns.(*TipSetExecutor).ExecuteTipSet
        /home/me/git/lotus/chain/consensus/filcns/compute_state.go:306
  - failed to get messages for block:
    github.com/filecoin-project/lotus/chain/store.(*ChainStore).BlockMsgsForTipset
        /home/me/git/lotus/chain/store/messages.go:147
  - failed to load msgmeta (bafy2bzacecb6oqlu7x3pufe6byqvpiwdo3vis2jguwhlr54ch6q665qs4bv7w):
    github.com/filecoin-project/lotus/chain/store.(*ChainStore).ReadMsgMetaCids
        /home/me/git/lotus/chain/store/messages.go:219
  - ipld: could not find bafy2bzacecb6oqlu7x3pufe6byqvpiwdo3vis2jguwhlr54ch6q665qs4bv7w
        * computing base fee:
    github.com/filecoin-project/lotus/chain/consensus/filcns.(*FilecoinEC).ValidateBlock.func3
        /home/me/git/lotus/chain/consensus/filcns/filecoin.go:149
  - error getting messages for: bafy2bzacect4widcroebz4ijcul3qew5hn6xif4durayol4hbtb2dq4ysedk4:
    github.com/filecoin-project/lotus/chain/store.(*ChainStore).ComputeBaseFee
        /home/me/git/lotus/chain/store/basefee.go:65
  - failed to load msgmeta (bafy2bzacecb6oqlu7x3pufe6byqvpiwdo3vis2jguwhlr54ch6q665qs4bv7w):
    github.com/filecoin-project/lotus/chain/store.(*ChainStore).ReadMsgMetaCids
        /home/me/git/lotus/chain/store/messages.go:219
  - ipld: could not find bafy2bzacecb6oqlu7x3pufe6byqvpiwdo3vis2jguwhlr54ch6q665qs4bv7w
        * get tipsetstate(102531, [bafy2bzacect4widcroebz4ijcul3qew5hn6xif4durayol4hbtb2dq4ysedk4 bafy2bzaceculwegadcv5pkcxqsfb4dhvg2dskxschgj5vu4ie5xolcxbdwsas bafy2bzacebwb6m4sgote2ydrrg2emep4mb3fu5k3epmumqaw2vgzvvnnagak6 bafy2bzaceajc5o53p22xr45tkkrrqicp2q6qnmvg5cq2hzzgm773wdwmqhn5u bafy2bzacechzactcbrsfddnv4q7hroklshedxx673k5d6l3zrbrunlvkdsaig bafy2bzaceat4dyrflcveuybogucth45s25ya5uv5tjkeysdg4qbe6ey3imngs]) failed:
    github.com/filecoin-project/lotus/chain/consensus/filcns.(*FilecoinEC).ValidateBlock.func4
        /home/me/git/lotus/chain/consensus/filcns/filecoin.go:170
  - getting block messages for tipset:
    github.com/filecoin-project/lotus/chain/consensus/filcns.(*TipSetExecutor).ExecuteTipSet
        /home/me/git/lotus/chain/consensus/filcns/compute_state.go:306
  - failed to get messages for block:
    github.com/filecoin-project/lotus/chain/store.(*ChainStore).BlockMsgsForTipset
        /home/me/git/lotus/chain/store/messages.go:147
  - failed to load msgmeta (bafy2bzacecb6oqlu7x3pufe6byqvpiwdo3vis2jguwhlr54ch6q665qs4bv7w):
    github.com/filecoin-project/lotus/chain/store.(*ChainStore).ReadMsgMetaCids
        /home/me/git/lotus/chain/store/messages.go:219
  - ipld: could not find bafy2bzacecb6oqlu7x3pufe6byqvpiwdo3vis2jguwhlr54ch6q665qs4bv7w
```

--- a/blockchain/chain/src/store/chain_store.rs
+++ b/blockchain/chain/src/store/chain_store.rs
@@ -595,6 +595,10 @@ where
                 }
             }
 
+            if h.epoch() > incl_roots_epoch {
+                recurse_links(&mut seen, *h.messages(), &mut load_block).await?;
+            }
+
             if h.epoch() > 0 {
                 for p in h.parents().cids() {
                     blocks_to_walk.push_back(*p);

